hardware-security-module/README.md

```diff
@@ -0,0 +1,127 @@
+
+1. install openssl 1.0.0+ or botan 1.10.0+
+2. download the source code from <https://dist.opendnssec.org/source/softhsm-2.5.0.tar.gz>
+3. `tar -xvf softhsm-2.5.0.tar.gz`
+4. `cd softhsm-2.5.0`
+5. `./configure --disable-gost` (would require additional libraries, turn it off unless you need 'gost' algorithm support for the Russian market)
+6. `make`
+7. `sudo make install`
+
+### PKCS#11 enabled fabric-ca-client binary
+To be able to register and enroll identities using an HSM you need a PKCS#11 enabled version of `fabric-ca-client`
+To install this use the following command
+
+```bash
+go install -tags 'pkcs11' github.com/hyperledger/fabric-ca/cmd/fabric-ca-client@latest
+```
+
+## Running the sample
+
+The Fabric test network is used to deploy and run this sample. Follow these steps in order:
+
+1. Create the test network and a channel (from the `test-network` folder).
+   ```
+   ./network.sh up createChannel -ca
+   ```
+
+2. Deploy one of the smart contract implementations (from the `test-network` folder).
+   ```
+   # To deploy the TypeScript chaincode implementation
+   ./network.sh deployCC -ccn basic -ccp ../asset-transfer-basic/chaincode-typescript/ -ccl typescript
+
+   # To deploy the Go chaincode implementation
+   ./network.sh deployCC -ccn basic -ccp ../asset-transfer-basic/chaincode-go/ -ccl go
+
+   # To deploy the Java chaincode implementation
+   ./network.sh deployCC -ccn basic -ccp ../asset-transfer-basic/chaincode-java/ -ccl java
+   ```
+
+## Initialize a token to store keys in SoftHSM
+
+If you have not initialized a token previously (or it has been deleted) then you will need to perform this one time operation
+
+```bash
+echo directories.tokendir = /tmp > $HOME/softhsm2.conf
+export SOFTHSM2_CONF=$HOME/softhsm2.conf
+softhsm2-util --init-token --slot 0 --label "ForFabric" --pin 98765432 --so-pin 1234
+```
+
+This will create a SoftHSM configuration file called `softhsm2.conf` and will be stored in the home directory. This is
+where the sample expects to find a SoftHSM configuration file
+
+The Security Officer PIN, specified with the `--so-pin` flag, can be used to re-initialize the token,
+and the user PIN (see below), specified with the `--pin` flag, is used by applications to access the token for
+generating and retrieving keys.
+
+## Enroll the HSM User
+
+A user, `HSMUser`, who is HSM managed needs to be registered then enrolled for the sample.
+
+If the "standard" PKCS11 library locations checked for by the script don't include the library(libsofthsm2.so) location for your environment set the `PKCS11_LIB` environment variable to define the library location.
+
+```bash
+PKCS11_LIB='<path to PKCS11 library location>' scripts/generate-hsm-user.sh HSMUser
+```
+
+This will register a user `HSMUser` with the CA in Org1 (if not already registered) and then enroll that user which will
+generate a certificate on the file system for use by the sample. The private key is stored in SoftHSM
```

Comment:
Seeing setting the `PKCS11_LIB` environment mentioned multiple times and multiple different ways in this document, I suspect it would be simpler to just mention it once here and have them set the environment variable for the shell so that it applies to all the following commands run in the shell. Then no need to mention the environment variable in the sections describing how to run each language version of the sample.
```suggestion
If your PKCS11 library (libsofthsm2.so) is not located in one of the typical Linux system locations checked by this sample's scripts and applications, you will need to explicitly specify the library location using the `PKCS11_LIB` environment variable.

```bash
export PKCS11_LIB='<path to PKCS11 library location>'```

Register a user `HSMUser` with the CA in Org1 (if not already registered) and then enroll that user which will
generate a certificate on the file system for use by the sample. The private key is stored in SoftHSM

```bash
scripts/generate-hsm-user.sh HSMUser```
```